framework/src/headers/ip/v6/icmp/router_advertisement.rs

```diff
@@ -0,0 +1,320 @@
+                     outgoing interface when sending traffic to its
+                     neighbors.
+      Future versions of this protocol may define new option types.
+      Receivers MUST silently ignore any options they do not recognize
+      and continue processing the message.
+
+*/
+
+const MANAGED_CFG_ADDR_POS: u8 = 0;
+const OTHER_CFG_POS: u8 = 1;
+
+#[derive(Debug)]
+#[repr(C, packed)]
+pub struct Icmpv6RouterAdvertisement<T>
+where
+    T: Ipv6VarHeader,
+{
+    icmp: Icmpv6Header<T>,
+    current_hop_limit: u8,
+    reserved_flags: u8,
+    router_lifetime: u16,
+    reachable_time: u32,
+    retrans_timer: u32,
+    options: u8,
+    _parent: PhantomData<T>,
+}
+
+impl<T> Default for Icmpv6RouterAdvertisement<T>
+where
+    T: Ipv6VarHeader,
+{
+    fn default() -> Icmpv6RouterAdvertisement<T> {
+        Icmpv6RouterAdvertisement {
+            icmp: Icmpv6Header {
+                msg_type: IcmpMessageType::RouterAdvertisement as u8,
+                code: 0,
+                checksum: 0,
+                ..Default::default()
+            },
+            current_hop_limit: 0,
+            reserved_flags: 0,
+            router_lifetime: 0,
+            reachable_time: 0,
+            retrans_timer: 0,
+            options: 0,
+            _parent: PhantomData,
+        }
+    }
+}
+
+impl<T> fmt::Display for Icmpv6RouterAdvertisement<T>
+where
+    T: Ipv6VarHeader,
+{
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(
+            f,
+            "msg_type: {} code: {} checksum: {}, current_hop_limit {}, reserved_flags {}, router_lifetime {}, reachable_time {}, retrans_timers {}",
+            self.msg_type().unwrap(),
+            self.code(),
+            self.checksum(),
+            self.current_hop_limit(),
+            self.reserved_flags(),
+            self.router_lifetime(),
+            self.reachable_time(),
+            self.retrans_timer(),
+        )
+    }
+}
+
+impl<T> EndOffset for Icmpv6RouterAdvertisement<T>
+where
+    T: Ipv6VarHeader,
+{
+    type PreviousHeader = T;
+
+    #[inline]
+    fn offset(&self) -> usize {
+        // ICMPv6 Header for Router Advertisement (Type + Code + Checksum)
+        // is always 8 bytes: (8 + 8 + 16) / 8 = 4
+        // Options are a variable length and will be manually parsed
+        16
+    }
+
+    #[inline]
+    fn size() -> usize {
+        // ICMPv6 Header is always 8 bytes so size = offset
+        16
+    }
+
+    #[inline]
+    fn payload_size(&self, hint: usize) -> usize {
+        // There is no payload size in the ICMPv6 header
+        hint - self.offset()
+    }
+
+    #[inline]
+    fn check_correct(&self, _prev: &T) -> bool {
+        true
+    }
+}
+
+impl<T> Icmpv6RouterAdvertisement<T>
+where
+    T: Ipv6VarHeader,
+{
+    pub fn get_source_link_layer_address_option(&self, payload_length: u16) -> Option<&MacAddress> {
+        let r = find_ndp_option(self, payload_length, NDPOptionType::SourceLinkLayerAddress);
```

Comment:
would love if `find_ndp_option` allowed us to pass in a type parameter that we expected to be returned instead of having the odd pattern match down below

Reply:
it would be nice to do the parsing up front in the packet rather than having to parse the options list each time. if we need > 1 options in a router advertisement, doesnt that mean we would have to parse the options N times?